README.md
# Vue2 Smooth Scroll
<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
[![All Contributors](https://img.shields.io/badge/all_contributors-2-orange.svg?style=flat-square)](#contributors-)
<!-- ALL-CONTRIBUTORS-BADGE:END -->

[![Vue 2.x](https://img.shields.io/badge/Vue-2.x-brightgreen.svg)](https://vuejs.org/v2/guide/)
[![npm](https://img.shields.io/npm/v/vue2-smooth-scroll.svg)](https://www.npmjs.com/package/vue2-smooth-scroll)
[![npm-downloades](https://img.shields.io/npm/dm/vue2-smooth-scroll.svg)](https://www.npmjs.com/package/vue2-smooth-scroll)
[![license](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/Yuliang-Lee/vue2-smooth-scroll/blob/master/LICENSE)

Lightweight Vue plugin for smooth-scrolling extended from [vue-smooth-scroll](https://github.com/alamcordeiro/vue-smooth-scroll).

For simple use-cases, the native [`scroll-behavior` CSS property](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-behavior) (working draft) may be enough.

**This package compatible with Vue 2.x and Vue 1.x**

**New package for Vue 3.x use this package [v-smooth-scroll](https://github.com/Yuliang-Lee/v-smooth-scroll)**

## Features

- Directive and programmatic API with global and local config overrides
- SSR
- Smooth, non-blocking animation using `requestAnimationFrame` (with fallback)
- Y-axis or vertical scrolling
- Specific scroll containers
- 1.3kB gzipped, 2.9kB min


## Installation
``` bash
npm install --save vue2-smooth-scroll
```

``` js
import VueSmoothScroll from 'vue2-smooth-scroll'
Vue.use(VueSmoothScroll)
```


## Usage
### Directive
``` html
<a href="#sec-3" v-smooth-scroll>Section 3</a>

<section id="sec-3"></section>
```

### Programmatic
``` js
const myEl = this.$refs.myEl || this.$el || document.getElementById(...)

this.$smoothScroll({
  scrollTo: myEl,
  hash: '#sampleHash'  // required if updateHistory is true
})
```

### Direct in `<script>`
``` html
<body>
  <div id="app">
    <a href="#bottom" v-smooth-scroll>Let's go to #bottom!</a>
    <div style="height: 2000px;"></div>
    <span id="bottom">bottom</span>
  </div>

  <script src="https://unpkg.com/vue/dist/vue.js"></script>
  <script src="https://unpkg.com/vue2-smooth-scroll"></script>
  <script>
    var app = new Vue({
      el: '#app'
    })
  </script>
</body>
```


## Custom options
### Defaults
``` js
{
  duration: 500,       // animation duration in ms
  offset: 0,           // offset in px from scroll element, can be positive or negative
  container: '',       // selector string or Element for scroll container, default is window
  updateHistory: true  // whether to push hash to history
  easingFunction: null // gives the ability to provide a custom easing function `t => ...`
                       // (see https://gist.github.com/gre/1650294 for examples)
                       // if nothing is given, it will defaults to `easeInOutCubic`
}
```

### Global
``` js
import VueSmoothScroll from 'vue2-smooth-scroll'

Vue.use(VueSmoothScroll, {
  duration: 400,
  updateHistory: false,
})
```

### Directive:
``` html
<div id="container">
  <a href="#div-id" v-smooth-scroll="{ duration: 1000, offset: -50, container: '#container' }">Anchor</a>

  <div id="div-id"></div>
</div>
```

### Programmatic
``` js
this.$smoothScroll({
  scrollTo: this.$refs.myEl,
  duration: 1000,
  offset: -50,
})
```


## License

[MIT](./LICENSE)


## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://github.com/jwhitmarsh"><img src="https://avatars2.githubusercontent.com/u/8026009?v=4" width="100px;" alt=""/><br /><sub><b>James</b></sub></a><br /><a href="https://github.com/Yuliang-Lee/vue2-smooth-scroll/commits?author=jwhitmarsh" title="Code">💻</a></td>
    <td align="center"><a href="https://github.com/a-kriya"><img src="https://avatars3.githubusercontent.com/u/26761352?v=4" width="100px;" alt=""/><br /><sub><b>a-kriya</b></sub></a><br /><a href="https://github.com/Yuliang-Lee/vue2-smooth-scroll/commits?author=a-kriya" title="Code">💻</a> <a href="https://github.com/Yuliang-Lee/vue2-smooth-scroll/issues?q=author%3Aa-kriya" title="Bug reports">🐛</a> <a href="https://github.com/Yuliang-Lee/vue2-smooth-scroll/commits?author=a-kriya" title="Documentation">📖</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!


History.md
v1.5.1 / 2021-08-31
==================

  * fix(package.json): 修复 main 入口错误指向 src 文件的问题
  * deps: update package-lock
  * Merge pull request #41 from Yuliang-Lee/dependabot/npm_and_yarn/browserslist-4.16.6
  * Merge pull request #40 from Yuliang-Lee/dependabot/npm_and_yarn/lodash-4.17.21
  * Merge pull request #39 from Yuliang-Lee/dependabot/npm_and_yarn/ssri-6.0.2
  * Merge pull request #38 from Yuliang-Lee/dependabot/npm_and_yarn/y18n-4.0.1
  * Merge pull request #37 from Yuliang-Lee/dependabot/npm_and_yarn/elliptic-6.5.4
  * Merge pull request #34 from Yuliang-Lee/dependabot/npm_and_yarn/ini-1.3.7
  * Merge pull request #33 from Yuliang-Lee/dependabot/npm_and_yarn/yargs-parser-13.1.2
  * Merge pull request #45 from Yuliang-Lee/dependabot/npm_and_yarn/path-parse-1.0.7
  * docs: update
  * docs: 修改readme 指向新的包
  * Merge pull request #29 from Yuliang-Lee/dependabot/npm_and_yarn/lodash-4.17.19

v1.5.0 / 2020-07-11
==================

  * feat: 增加easingFunction, 增加 husky 做提交检查
  * Merge pull request #28 from qnp/feature/custom-easing-function
  * option to provide custom easing function
  * Merge pull request #25 from a-kriya/patch-1
  * build
  * Merge pull request #27 from jwhitmarsh/fix/#26
  * fix(index): first cut at fixing #26
  * Update README to incorporate recent changes

v1.4.2 / 2020-06-18
==================

  * fix(index): first cut at fixing #26

v1.4.0 / 2020-06-16
==================

  * build(types): add TypeScript types
  * feat: export smoothScroll() method so it can be used programatically

v1.3.0 / 2020-04-13
==================

  * deps: update
  * fix: Set output.globalObject to "this" to fix SSR (https://github.com/Yuliang-Lee/vue2-smooth-scroll/pull/19)

v1.2.0 / 2020-01-18
==================

  * feat: use in component
  * Document updateHistory options (#13)

v1.1.1 / 2019-11-02
==================

  * deps: update other deps
  * build(deps): bump eslint-utils from 1.3.1 to 1.4.3

v1.1.0 / 2019-06-25
==================

  * deps: update
  * feat: add config option to enable/disable updating window.history https://github.com/Yuliang-Lee/vue2-smooth-scroll/pull/9

v1.0.5 / 2019-02-16
==================

  * deps: update dependencies
  * fix: there is no need to explicit register plugin

v1.0.4 / 2019-02-08
==================

  * fix: SSR

v1.0.3 / 2018-04-24
==================

  * fix: remove object.assign compatible IE


src/index.js
// we use requestAnimationFrame to be called by the browser before every repaint
let requestAnimationFrame;

// Get the top position of an element in the document
const getTop = function(element, start) {
  // return value of html.getBoundingClientRect().top ... IE : 0, other browsers : -pageYOffset
  if (element.nodeName === 'HTML') return -start;

  const elementTop = element.getBoundingClientRect().top
  const scrollMarginTop = parseInt(window.getComputedStyle(element).scrollMarginTop || 0, 10)
  return elementTop + start - scrollMarginTop;
};

function getDefaultConfig() {
  return {
    duration: 500,
    offset: 0,
    container: window,
    updateHistory: true,
    easingFunction: null,
  };
}

const smoothScrollCtx = Symbol('smoothScrollCtx')

function _smoothScroll({ scrollTo, offset, duration, container, updateHistory, hash, easingFunction }) {
  if (!requestAnimationFrame) {
    requestAnimationFrame = window.requestAnimationFrame ||
    function(fn) {
      window.setTimeout(fn, 16);
    };
  }

  // Using the history api to solve issue: back doesn't work
  // most browser don't update :target when the history api is used:
  // THIS IS A BUG FROM THE BROWSERS.
  if (updateHistory && window.history.pushState && location.hash !== hash) window.history.pushState('', '', hash);


  const startPoint = container.scrollTop || window.pageYOffset;
  // Get the top position of an element in the document
  // return value of html.getBoundingClientRect().top ... IE : 0, other browsers : -pageYOffset
  let end = getTop(scrollTo, startPoint);

  // Ajusts offset from the end
  end += offset;

  const clock = Date.now();
  const step = function() {
    // the time elapsed from the beginning of the scroll
    const elapsed = Date.now() - clock;
    // calculate the scroll position we should be in
    let position = end;
    let easeFn = (typeof easingFunction === 'function' && easingFunction) || easeInOutCubic;
    if (elapsed < duration) {
      position = startPoint + (end - startPoint) * easeFn(elapsed / duration);

      requestAnimationFrame(step);
    } else if (updateHistory) {
      location.replace('#' + scrollTo.id);
      // this will cause the :target to be activated.
    }

    container === window ? container.scrollTo(0, position) : (container.scrollTop = position);
  };
  step();
}

const VueSmoothScroll = {
  install(Vue, config) {
    Vue.directive('smooth-scroll', {
      inserted(el, binding, vnode) {
        // Do not initialize smoothScroll when running server side, handle it in client
        // We do not want this script to be applied in browsers that do not support those
        // That means no smoothscroll on IE9 and below.
        if (typeof window !== 'object' || window.pageYOffset === undefined) return;

        let resolvedConfig = Object.assign({}, getDefaultConfig());
        if (config) {
          Object.assign(resolvedConfig, config);
        }

        let { duration, offset, container, updateHistory, easingFunction } = binding.value || {};
        duration = duration || resolvedConfig.duration;
        offset = offset || resolvedConfig.offset;
        container = container || resolvedConfig.container;
        updateHistory = updateHistory !== undefined ? updateHistory : resolvedConfig.updateHistory;
        easingFunction = easingFunction || resolvedConfig.easingFunction;

        if (typeof container === 'string') {
          container = document.querySelector(container);
        }

        const clickHandler = function(ev) {
          ev.preventDefault();
          const hash = vnode.data.attrs.href;
          const scrollTo = document.getElementById(hash.substring(1));
          if (!scrollTo) return; // Do not scroll to non-existing node

          _smoothScroll({ scrollTo, offset, duration, container, updateHistory, hash, easingFunction });
        }
        // Attach the smoothscroll function
        el.addEventListener('click', clickHandler);

        el[smoothScrollCtx] = {
          clickHandler
        }
      },
      unbind(el) {
        el.removeEventListener('click', el[smoothScrollCtx].clickHandler)
        el[smoothScrollCtx] = null
      }
    });

    Vue.prototype.$smoothScroll = (args) => {
      const resolvedArgs = Object.assign({}, getDefaultConfig(), config, args);
      return _smoothScroll(resolvedArgs)
    }
  }
};

export default VueSmoothScroll

/* istanbul ignore if */
if (typeof window !== 'undefined' && window.Vue) {
  window.Vue.use(VueSmoothScroll)
}

/**
 * ease in out function
 * @see https://gist.github.com/gre/1650294
 */
function easeInOutCubic(t) {
  return t < 0.5 ? 4 * t * t * t : (t - 1) * (2 * t - 2) * (2 * t - 2) + 1;
}
